File: kijung/day1_1.java
//http://boj.kr/8b513be94bd74fa1aa717a7fcad9a8aa

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.io.IOException;

import java.lang.StringBuilder;

import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());
        int[] arr = new int[M];
        StringBuilder sb = new StringBuilder();
        method(1, 0, N, M, arr, sb);

        bw.write(sb.toString());
        bw.flush();
        bw.close();
    }

    public static void method(int at, int depth, int N, int M, int[] arr, StringBuilder sb) {
        if (M == depth) {
            for (int value : arr) {
                sb.append(value).append(" ");
            }
            sb.append("\n");
            return;
        }

        for (int i = at; i <= N; i++) {
            arr[depth] = i;
            method(i, depth + 1, N, M, arr, sb);
        }
    }
}

File: yeonjin/day1_1.kt
/*
 * [문제 요약]
 * 1부터 N까지 자연수 중에서 M개를 고른 수열을 모두 출력
 * 같은 수를 여러 번 골라도 되고(중복 허용), 수열은 '비내림차순'.
 * (즉, a1 ≤ a2 ≤ ... ≤ aM)
 *
 * [입력]
 * N M (1 ≤ M ≤ N ≤ 8)
 *
 * [출력]
 * 조건을 만족하는 모든 수열을 한 줄에 하나씩 공백으로 구분하여 출력.
 */

import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.StringTokenizer

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val st = StringTokenizer(br.readLine())
    val n = st.nextToken().toInt()
    val m = st.nextToken().toInt()

    val seq = IntArray(m)
    val result = StringBuilder()

    fun backtrack(start: Int, depth: Int) {
        if (depth == m) {
            result.append(seq.joinToString(" "))
            result.append('\n')
            return
        }
        for (i in start..n) {
            seq[depth] = i
            backtrack(i, depth + 1)
        }
    }

    backtrack(1, 0)
    print(result)
}